File: src/cpp_package/src/hw2.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include <memory>

using namespace std::chrono_literals;

class NumberCounter : public rclcpp::Node
{
public:
    NumberCounter() : Node("number_counter")
    {
        RCLCPP_INFO(this->get_logger(), "number_counter is created... It publishes and subscribes");

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number",
            10,
            std::bind(&NumberCounter::callbackSubscription, this, std::placeholders::_1)
        );

        timer_ = this->create_wall_timer(1s, std::bind(&NumberCounter::publishNumber, this));
    }

private:
    void publishNumber()
    {
        auto message = example_interfaces::msg::Int64();
        message.data = counter_;
        RCLCPP_INFO(this->get_logger(), "Publish the number %ld from NumberCounter", counter_);
        publisher_->publish(message);
    }

    void callbackSubscription(const example_interfaces::msg::Int64::SharedPtr receivedMessage)
    {
        RCLCPP_INFO(this->get_logger(), "Received number is: %ld", receivedMessage->data);
        counter_++;
    }

    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::TimerBase::SharedPtr timer_;
    int64_t counter_{0};
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/python_package/python_package/hw2.py
# This is the second node
# This file has 1 publisher and 1 subscriber
# Name: number_counter
# This will subscribe to a topic named 'number'
# And will publish under a topic named 'number_count'
# The class will also have a server named 'reset_counter'
# If the client requests True, then the counter will be 0

import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
from numpy.random import randint

class HW2(Node):
    def __init__(self):
        super().__init__("number_counter")

        self.get_logger().info("number_counter is created... It published and subscribes")
        self.counter_ = 0
        self.the_publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.the_subscriber_ = self.create_subscription(Int64, "number", self.callback_subscription, 10)

        self.create_timer(1.0 , self.publish_number)


        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_counter_reset)
    
    def callback_counter_reset(self, request, response):
        if request.data:
            self.counter_ = 0
            self.get_logger().info("Counter has been reset.")
            response.success = True
            response.message = "The success has been set to True bro... The counter has been reset"
        else:
            response.success = False
            response.message = "TNO RESET"

        return response

    def publish_number(self):
        the_message = Int64()
        the_message.data = self.counter_
        self.get_logger().info(f"Publish the number {self.counter_} from HW2")
        self.the_publisher_.publish(the_message)
          
    def callback_subscription(self, received_message):
        self.get_logger().info(f"Received number is: {received_message.data}")
        self.counter_ += 1


def main(args=None):
    rclpy.init(args=args)
    theNode = HW2()
    rclpy.spin(theNode)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
